controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -372,32 +379,33 @@ public boolean checkAuthorization(String resource, AuthHandler.Permissions expec
                 allowedLevel = currentInterceptor.authorize(resource);
             }
             if (allowedLevel.ordinal() < expectedLevel.ordinal()) {
-                return false;
+                throw new AuthorizationException("Access not allowed");
             }
-            return true;
-        } else {
-            return true;
         }
+        return "";
     }
 
-    public boolean checkAuthorizationWithToken(String resource, AuthHandler.Permissions expectedLevel) {
+    private String checkAuthorizationWithToken(String resource, AuthHandler.Permissions expectedLevel) throws AuthorizationException {
         if (isAuthEnabled) {
-            boolean retVal = checkAuthorization(resource, expectedLevel);
-            if (retVal) {
-                PravegaInterceptor interceptor = PravegaInterceptor.INTERCEPTOR_OBJECT.get();
-                interceptor.setDelegationToken(resource, expectedLevel, tokenSigningKey);
-            }
-            return retVal;
-        } else {
-            return true;
+            checkAuthorization(resource, expectedLevel);
+            return createDelegationToken(resource, expectedLevel, tokenSigningKey);
         }
+        return "";
     }
 
-    private String getCurrentDelegationToken() {
+    private String createDelegationToken(String resource, AuthHandler.Permissions expectedLevel, String tokenSigningKey) {
         if (isAuthEnabled) {
-            return PravegaInterceptor.retrieveDelegationToken(tokenSigningKey);
-        } else {
-            return "";
+            Map<String, Object> claims = new HashMap<>();
+
+            claims.put(resource, String.valueOf(expectedLevel));
+
+            return Jwts.builder()
+                       .setSubject("segmentstoreresource")
```

Comment:
Why is the subject `segmentstoreresource`?

Reply:
This is the meaning of subject tag for JWT. https://tools.ietf.org/html/rfc7519#section-4.1.2
My intention behind naming it `segmentstoreresource` was that rest of the claims are related to this. Do share if you have any other suggestions for this.

Reply:
To expand on this, the subject field of a JWT typically contains a username, because the claims of the token are in regards to a user.   For identity tokens, the claims are about the subject's identity (e.g. their email address, user id).   For access tokens, the claims are about the subject's permissions w.r.t a particular resource.   For example:
"Subject 'eron' has 'read permission for X' according to issuer 'Pravega Controller'.  The audience for this information is the 'Pravega Segment Store' ."

A better use for the subject field in this case would be the username of the authenticated user, for logging purposes in the segment store.